Added command to get diff with Exodus platform

nautikos1235 opened the pull request:
Related to https://github.com/Exodus-Privacy/etip/issues/17

Adds a `compare_with_exodus` command that gets trackers from the Exodus API, then compare them with the local trackers database having `is_in_exodus = True`

Running it on a fresh install with trackers, it logs:
```
Retrieved 195 trackers from Exodus
Found 195 trackers in ETIP DB expected to be in Exodus
[...]
Lookup results:
** FOUND_AND_IDENTICAL: 147
** FOUND_BUT_DIFFERENT: 0
** MULTIPLE_MATCHES_FOUND_IN_ETIP: 48
** NOT_FOUND_IN_ETIP: 0
```

WIP because:
- [x] I'll add some more tests soon and refactor them a bit
- [x] Update doc/readme
- [x] Question (@simpnu?): is it ok for you to do case-insensitive compare ?
- [x] Need to `--autosquash` PR

But you can start reviewing & commenting @simpnu :)
Especially the requirements part, in which I might be pushing too much stuff...

simpnu commented on etip/trackers/tests.py:
```diff
@@ -1,6 +1,9 @@
 from .models import Tracker
-from io import BytesIO
+from io import BytesIO, StringIO
+from mock import patch
```
Replace this by `from unittest.mock import patch` and you won't need to add mock (and its dependencies) to ETIP requirements.

nautikos1235 replied:
Done; removed 3 deps!

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,95 @@
+    def handle(self, *args, **options):
+        exodus_trackers = self.get_all_from_exodus()
+        self.stdout.write(
+            'Retrieved ' + str(len(exodus_trackers)) + ' tracker from Exodus '
```
Please use format for printing logs, example here:
`'Retrieved {} trackers from Exodus'.format(len(exodus_trackers))`

See https://pyformat.info/

nautikos1235 replied:
Thanks, updated

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,99 @@
+            raise CommandError('Empty response')
+        return resp.get('trackers')
+
+    def count_from_etip(self):
```
The point of this function is only to print the number of trackers?
If so, I would:
- remove the unnecessary "return" at the end
- do `in_exodus_trackers_nb = Tracker.objects.filter(is_in_exodus=True).count()`

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,99 @@
+
+    def handle(self, *args, **options):
+        exodus_trackers = self.get_all_from_exodus()
+        self.stdout.write(
```
I would add the log line in the get_all_from_exodus function for better readability of the "handle" function.

nautikos1235 replied:
I don't agree; The `get_all_from_exodus` is my service, while logging to stdout is rendering.
This way of rendering is specific to building an admin command and would not be relevant if building a web interface (while calling the api will always be). I know this separation of concerns is not enforced in all the script, but I would keep it where it's right :D
WDYT ?

simpnu replied:
Fair enough!

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,99 @@
+        self.count_from_etip()
+        self.lookup_trackers(exodus_trackers)
+
+    def __exact_lookup(self, tracker_details):
```
I think maybe adding docstrings for this function (maybe for the others too) to make it clearer what it does and returns.

nautikos1235 replied:
The `__exact_lookup` method was badly named, had too many responsabilities and was overly complicated =) Refactored it into `find_etip_tracker`, should be clearer. Please tell me if you think it still needs docstrings!

["Don't document bad code; rewrite it"](https://twitter.com/voidspace/status/252783747785687040) :p

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,99 @@
+                return (LOOKUP_NOT_FOUND, None)
+            elif not results:
+                continue
+            elif len(results) != 1:
```
results.count() is probably preferable

nautikos1235 replied:
Thanks for the tip, taken it into account (but also refactored the whole function ^^)

nautikos1235 commented:
@simpnu Sorry I am not familiar with the GH review system and can't comment on your "outdated" comments ^^

> I think maybe adding docstrings for this function (maybe for the others too) to make it clearer what it does and returns.

This function was not understandable and overly complicated =) Refactored it, should be clearer. Please tell me if you think it still needs docstrings! "Don't document bad code; rewrite it" :p

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
         self.assertJSONEqual(f.getvalue(), expected_json)
+
+
+EXODUS_API_BASE_URL = 'https://reports.exodus-privacy.eu.org'
```
I would either put the constants at the beginning of the file (if we think it can be useful later one for other classes) or inside the class.

nautikos1235 replied:
Agreed, moved it!

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
+            with self.assertRaises(CommandError) as e:
+                out = StringIO()
+                call_command('compare_with_exodus', stdout=out)
+            error_msg = str(e.exception)
```
Indent this line and the following one to the left I think. Same goes for other tests (no need to do the assertion inside the with patch xxx)

nautikos1235 replied:
Moved it!

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
+    def test_api_gets_called_correctly(self):
+        with patch('requests.get') as mocked_get:
+            mocked_get.return_value.status_code = 200
+            out = StringIO()
```
I'm wondering if we need to keep this line if we don't make any assertion of the output of the command.

nautikos1235 replied:
Defining an `stdout=` param in the `call_command` is useful to avoid having logs spit out in the console when executing tests (it overrides standard stdout), but it's true I don't need a named variable if I don't use it after. Corrected!

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
+            Tracker(name="Not In Exodus", is_in_exodus=False).save()
+            out = StringIO()
+            call_command('compare_with_exodus', stdout=out)
+            self.assertIn(
```
This can fit on one line

nautikos1235 replied:
Fixed it!

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
+    def test_select_only_in_exodus_trackers(self):
+        with patch('requests.get') as mocked_get:
+            mocked_get.return_value.status_code = 200
+            Tracker(name="In Exodus", is_in_exodus=True).save()
```
I would move the Trackers creation:
- outside of the mock 
- in the beginning of the test function

nautikos1235 replied:
Done! Refactored & added missing tests

simpnu commented on etip/trackers/tests.py:
```diff
@@ -321,3 +324,155 @@ def test_with_trackers(self):
+                    '1': {
+                        'name': tracker_1.name,
+                        'code_signature': tracker_1.code_signature,
+                        'network_signature':
```
this can fit on one line too :)

nautikos1235 replied:
Fixed it

simpnu commented on etip/trackers/management/commands/compare_with_exodus.py:
```diff
@@ -0,0 +1,114 @@
+                nb_of_trackers
+            )
+        )
+        return
```
Still not a big fan of this return :)

nautikos1235 replied:
Yep, thought I had removed it ^^

simpnu commented on requirements.txt:
```diff
@@ -1,4 +1,17 @@
+pbr==5.1.3
+pkg-resources==0.0.0
+pycodestyle==2.5.0
+pyflakes==2.1.1
```
I'm not sure we want to add pyflakes (and its dependencies) here. Otherwise that means that it will be installed on the production as well.

nautikos1235 replied:
Yes I was not sure either; reverted it

nautikos1235 commented:
Oups, found a bug while refactoring & adding text (query filter AND instead of OR !).
Now I have 48 MULTIPLE_MATCHES_FOUND_IN_ETIP :D

> One more important thing: for more convenience, I think the script should show directly what differences it finds (and not only the name of the trackers which have differences). That could be an option to add or simply the default behavior. Do you think that's feasible?

I'll look into that, should not be too hard

nautikos1235 commented:
Looks like It's missing some rules to be relevant @simpnu : most duplicates are because of the "Network detection rule" or "website", while their "Code detection rule" is not the same.

An example with [all the facebook trackers](https://etip.exodus-privacy.eu.org/?tracker_name=facebook).

The script should probably **search** only by `name` or `code_signature` (I guess these are supposed to be unique), then **compare** with all fields (website...). Do you agree?

nautikos1235 commented:
@simpnu Updated !
- [x] Switched to case-sensitive compare
- [x] Search only by **name & code signature** (not network signature nor website)
- [x] Added full diff details
- [x] Added --quiet (-q) option to avoid full display
- [x] Added tests for the 2 command options (exodus hostname override & quiet)

Ready for merge!

simpnu commented:
Awesome! I removed 2 extra requirements which were not necessary and merged your PR, thanks a lot :).
